src/app/clusters/groups-server/groups-server.cpp

```diff
@@ -232,141 +261,137 @@ bool emberAfGroupsClusterViewGroupCallback(app::CommandHandler * commandObj, con
-                count++;
-            }
-        }
-    }
-    else
-    {
-        auto iter = groupList.begin();
-        while (iter.Next())
-        {
-            GroupId groupId = iter.GetValue();
-            emberAfGroupsClusterPrint(" 0x%02" PRIx16, groupId);
-            for (j = 0; j < EMBER_BINDING_TABLE_SIZE; j++)
-            {
-                EmberBindingTableEntry entry;
-                status = emberGetBinding(j, &entry);
-                if ((status == EMBER_SUCCESS) && (entry.type == EMBER_MULTICAST_BINDING))
+                auto iter = mCommandData.groupList.begin();
+                while (iter.Next())
                 {
-                    if (entry.local == emberAfCurrentEndpoint() && entry.groupId == groupId)
+                    if (mapping.group == iter.GetValue())
                     {
-                        list[listLen]     = EMBER_LOW_BYTE(groupId);
-                        list[listLen + 1] = EMBER_HIGH_BYTE(groupId);
-                        listLen           = static_cast<uint8_t>(listLen + 2);
-                        count++;
+                        ReturnErrorOnFailure(chip::app::DataModel::Encode(writer, TLV::AnonymousTag, mapping.group));
+                        matchCount++;
+                        emberAfGroupsClusterPrint(" 0x%02" PRIx16, mapping.group);
+                        break;
                     }
                 }
+                ReturnErrorOnFailure(iter.GetStatus());
             }
+            emberAfGroupsClusterPrintln("]");
+
+            // 1.3.6.3.1: If the GroupList field contains one or more group-id’s but does not contain any group
+            // of which the entity is a member, the entity SHALL only respond if the command is unicast
+            VerifyOrReturnError((0 == requestCount) || (matchCount > 0) ||
+                                    (emberAfCurrentCommand()->type == EMBER_INCOMING_UNICAST),
+                                CHIP_ERROR_INVALID_DATA_LIST);
+
+            ReturnErrorOnFailure(writer.EndContainer(type));
         }
-        err = iter.GetStatus();
+
+        ReturnErrorOnFailure(writer.EndContainer(outer));
+        return CHIP_NO_ERROR;
     }
+};
 
-    emberAfGroupsClusterPrintln("]");
+bool emberAfGroupsClusterGetGroupMembershipCallback(app::CommandHandler * commandObj, const app::ConcreteCommandPath & commandPath,
+                                                    const Commands::GetGroupMembership::DecodableType & commandData)
+{
+    auto fabricIndex   = GetFabricIndex(commandObj);
+    auto * groups      = GetGroupDataProvider();
+    EmberStatus status = EMBER_ZCL_STATUS_FAILURE;
+    CHIP_ERROR err     = CHIP_NO_ERROR;
 
-    if (err != CHIP_NO_ERROR)
+    if (nullptr == groups)
     {
-        status = emberAfSendImmediateDefaultResponse(EMBER_ZCL_STATUS_MALFORMED_COMMAND);
-        err    = CHIP_NO_ERROR;
-        goto exit;
+        status = emberAfSendImmediateDefaultResponse(EMBER_ZCL_STATUS_FAILURE);
```

Comment:
Should this be a failure, or a response with empty list?

Reply:
I think this is a failure, the group data provider should never be null.